jax/_src/test_util.py

```diff
@@ -931,10 +931,21 @@ def tearDownClass(cls):
   def rng(self):
     return self._rng
 
-  def assertArraysEqual(self, x, y, *, check_dtypes=True, err_msg=''):
+  def assertArraysEqual(self, x, y, *, check_dtypes=True, err_msg='', allow_object_dtype=False):
     """Assert that x and y arrays are exactly equal."""
     if check_dtypes:
       self.assertDtypesMatch(x, y)
+    x = np.asarray(x)
+    y = np.asarray(y)
+
+    if (not allow_object_dtype) and (x.dtype == object or y.dtype == object):
+      # See https://github.com/google/jax/issues/17867
+      raise TypeError(
+        "assertArraysEqual may be poorly behaved when np.asarray casts to dtype=object. "
+        "If comparing PRNG keys, consider random_test.KeyArrayTest.assertKeysEqual. "
```

Comment:
should AssertKeysEqual be in test_util too?

Reply:
It's not currently relevant anywhere outside of `random_test.py`, so I lean toward leaving it there until that changes.

Reply:
Ok. My point was that if someone wants to use it outside of random_test.py, then doing random_test.KeyArrayTest.assertKeysEqual is awkward.

Does it hurt to put in test_util.py? It's visibility controlled anyways. No strong opinions though from me.

Reply:
It's all internal – if someone wants to use it outside `random_test`, they can move it to `test_util`.